### ISSUE TITLE
N+1 Queries with nested Includes

### ISSUE THREAD

**Issue by lgebhardt** [JSONAPI-Resources/jsonapi-resources#206]
#204 and #205 address eager loading of resources and remove the N+1 query issue when resources are included. However the N+1 case still exists for nested includes. For example `include=author,author.comments` results in 19 queries in the test code.

~~I've made a branch to show this (https://github.com/cerebris/jsonapi-resources/tree/query_count). The tests in this work, but aren't what I want in the final codebase.~~ 
Edit: Branch deleted

**Comment by lgebhardt** [comment]
Here're the queries that are being run for fetching the posts with `include=author,author.comments`:

```
Processing by PostsController#index as HTML
  Parameters: {"include"=>"author,author.comments"}
  Post Load (0.2ms)  SELECT "posts".* FROM "posts"
  Person Load (0.2ms)  SELECT "people".* FROM "people" WHERE "people"."id" IN (1, 3, 4)
  Comment Load (0.1ms)  SELECT "comments".* FROM "comments" WHERE "comments"."author_id" IN (1, 3, 4)
  Comment Load (0.1ms)  SELECT "comments".* FROM "comments" WHERE "comments"."author_id" = ?  [["author_id", 1]]
  Comment Load (0.0ms)  SELECT "comments".* FROM "comments" WHERE "comments"."author_id" = ?  [["author_id", 1]]
  Comment Load (0.0ms)  SELECT "comments".* FROM "comments" WHERE "comments"."author_id" = ?  [["author_id", 3]]
  Comment Load (0.0ms)  SELECT "comments".* FROM "comments" WHERE "comments"."author_id" = ?  [["author_id", 3]]
  Comment Load (0.0ms)  SELECT "comments".* FROM "comments" WHERE "comments"."author_id" = ?  [["author_id", 3]]
  Comment Load (0.0ms)  SELECT "comments".* FROM "comments" WHERE "comments"."author_id" = ?  [["author_id", 3]]
  Comment Load (0.0ms)  SELECT "comments".* FROM "comments" WHERE "comments"."author_id" = ?  [["author_id", 3]]
  Comment Load (0.0ms)  SELECT "comments".* FROM "comments" WHERE "comments"."author_id" = ?  [["author_id", 3]]
  Comment Load (0.0ms)  SELECT "comments".* FROM "comments" WHERE "comments"."author_id" = ?  [["author_id", 3]]
  Comment Load (0.0ms)  SELECT "comments".* FROM "comments" WHERE "comments"."author_id" = ?  [["author_id", 3]]
  Comment Load (0.0ms)  SELECT "comments".* FROM "comments" WHERE "comments"."author_id" = ?  [["author_id", 1]]
  Comment Load (0.0ms)  SELECT "comments".* FROM "comments" WHERE "comments"."author_id" = ?  [["author_id", 4]]
  Comment Load (0.0ms)  SELECT "comments".* FROM "comments" WHERE "comments"."author_id" = ?  [["author_id", 4]]
  Comment Load (0.0ms)  SELECT "comments".* FROM "comments" WHERE "comments"."author_id" = ?  [["author_id", 3]]
  Comment Load (0.0ms)  SELECT "comments".* FROM "comments" WHERE "comments"."author_id" = ?  [["author_id", 3]]
  Comment Load (0.0ms)  SELECT "comments".* FROM "comments" WHERE "comments"."author_id" = ?  [["author_id", 3]]
Completed 200 OK in 53ms (Views: 4.3ms | ActiveRecord: 1.1ms)
```

It appears the eager loading is working based on the first three queries, however the comments aren't using the retrieved comment data when they are being instantiated.

**Comment by slaskis** [comment]
I didn't know about query_diet so instead I used instrumentation to track the queries:

```

## A little hack to get some numbers of how many queries was used
queries = {}
ActiveSupport::Notifications.subscribe('sql.active_record') do |name, start, finish, id, payload|
  name = payload[:name] || payload[:sql].gsub(/((?:FROM|JOIN) "?[\w_]+"?)/, "\1")
  queries[name] ||= 0
  queries[name] += 1
end
ActiveSupport::Notifications.subscribe('start_processing.action_controller') do |name, start, finish, id, payload|
  queries = {}
end
ActiveSupport::Notifications.subscribe('process_action.action_controller') do |name, start, finish, id, payload|
  Rails.logger.info("queries:")
  total = 0
  internal = 0
  queries.keys.sort.each do |name|
    count = queries[name]
    Rails.logger.info(" - #{name}: #{count}")
    if name == 'SCHEMA' || name == 'CACHE'
      internal += count
    end
    total += count
  end
  Rails.logger.info(" = #{total} (- #{internal} internal)")
end
```

I just logged it (as you can see) but maybe you can change it to something in a test helper?

**Comment by lgebhardt** [comment]
Thanks, I'll take a look at it. Query_diet was literally the first thing I found that worked for the task. I've hacked how it's used, and it's not very elegant. Still it beats looking only at the log file one test at a time.

**Comment by slaskis** [comment]
Cool, if you get it working feel free to update this PR and I'll try to debug on my end too

**Comment by lgebhardt** [comment]
I just updated #205 with a simple query_count method based on the `ActiveSupport::Notifications.subscribe('sql.active_record')` method. I'm going to get rid of the query_count branch now.

**Comment by slaskis** [comment]
Quick work! I'll have a look

**Comment by vrybas** [comment]
@lgebhardt @slaskis fixed by #205? Can be closed?

Doesn't seem like this is the issue anymore on 'master'

``` ruby
get '/articles?include=user,user.comments'
```

``` sql
Article Load (1.3ms)  SELECT "articles".* FROM "articles"  ORDER BY "articles"."id" ASC
User Load (0.6ms)     SELECT "users".* FROM "users" WHERE "users"."id" IN (1)
Comment Load (0.6ms)  SELECT "comments".* FROM "comments" WHERE "comments"."user_id" IN (1)
```
